### DISCUSSION TITLE
How to test pseudo elements?

### DISCUSSION THREAD

**Discussion by sashberd** [vitest-dev/vitest#6550] [Q&A]
I have next styled-component that I need to test:
```
export const StyledComponentLabel = styled.label<{
  isRequired: boolean
  isValid?: boolean
}>`
  color: ${({ isValid, state }) =>
    isValid === false  ? COLORS.redBase : COLORS.black};
  display: block;
  font-size: ${FONT_SIZES.sm};
  margin: 0;
  margin-bottom: ${SPACINGS.xxs};
  font-weight: ${FONT_WEIGHTS.bold};
  position: relative;
  white-space: pre;

  &::after {
    content: '*';
    opacity: ${({ isRequired }) => (isRequired ? 1 : 0)};
    margin-left: 5px;
    color: red;
  }
`
```

I was able to test the color of the component, but not the opacity of after pseudo element

I was trying next code:

```
 const label = screen.getByTestId(dataTestId.myComponent)
 const pseudoElement = window.getComputedStyle(label,'::after')
 expect(pseudoElement.opacity).toBe('1')
```
For both conditions, 0 and 1, the returned opacity value is an empty string. I've encountered discussions suggesting that JSDom in Vitest might not fully support pseudo-class testing. Is this accurate, or are there alternative approaches to achieve coverage for such scenarios?

**Comment by AriPerkkio** [comment]
Does with work properly in Browser Mode? No need to simulate any DOM APIs.

**Comment by sashberd** [comment]
@AriPerkkio Are you referring to [this](https://vitest.dev/guide/browser/). If so, I'm not using this approach in my project

**Reply by AriPerkkio** [reply]
Yep, here is working example: https://stackblitz.com/~/edit/vitest-pesudo-element-assertion?file=test/example.test.ts&file=vite.config.ts

```ts
test("pseudo elements", async () => {
    const div = document.getElementById("example")!

    const before = window.getComputedStyle(div,'::before')
    expect(before.opacity).toBe('0.75')

    const after = window.getComputedStyle(div,'::after')
    expect(after.opacity).toBe('0.123')
});
```
